File: rust/advent-of-code-2021/src/solutions/three.rs
use std::{
    fs::File,
    io::{BufRead, BufReader},
};

pub fn read_bits_from_file() -> Vec<String> {
    let file = File::open("src/inputs/three.txt").expect("Cannot find/open input file for day 3");
    let reader = BufReader::new(file);

    let mut result: Vec<String> = vec![];
    for line in reader.lines() {
        let current_line = line.expect("Cannot read line. Aborting...");
        result.push(current_line.trim().to_string());
    }
    result
}

pub fn process_bits(bits: &Vec<String>) -> isize {
    let mut bit_counter: Vec<isize> = vec![0; bits[0].len()];
    for bit_line in bits {
        let mut index = 0;
        for part in bit_line.split("") {
            if part == "" {
                continue;
            }

            bit_counter[index] += if part == "0" { -1 } else { 1 };
            index += 1;
        }
    }

    let mut gamma = String::from("");
    let mut epsilon = String::from("");
    for bit_count in bit_counter {
        if bit_count > 0 {
            gamma.push_str("1");
            epsilon.push_str("0");
        } else {
            gamma.push_str("0");
            epsilon.push_str("1");
        }
    }

    let gamma_value = isize::from_str_radix(gamma.as_str(), 2).unwrap();
    let epsilon_value = isize::from_str_radix(epsilon.as_str(), 2).unwrap();

    gamma_value * epsilon_value
}

pub fn find_oxygen_rating(bits: &Vec<String>) -> isize {
    let mut i = 0;

    let mut result = bits.clone();
    while result.len() > 1 {
        let mut counter = 0;

        for bit_line in &result {
            counter += if bit_line.chars().nth(i).unwrap() == '1' {
                1
            } else {
                -1
            };
        }

        let filter = if counter < 0 { '1' } else { '0' };

        result = result
            .iter()
            .filter(|bit| bit.chars().nth(i).unwrap().eq(&filter))
            .map(|bit| String::from(bit))
            .collect();
        i += 1;
    }

    isize::from_str_radix(result.get(0).unwrap(), 2).unwrap()
}

pub fn find_co2_rating(bits: &Vec<String>) -> isize {
    let mut i = 0;

    let mut result = bits.clone();
    while result.len() > 1 {
        let mut counter = 0;

        for bit_line in &result {
            counter += if bit_line.chars().nth(i).unwrap() == '1' {
                1
            } else {
                -1
            };
        }

        let filter = if counter < 0 { '0' } else { '1' };

        result = result
            .iter()
            .filter(|bit| bit.chars().nth(i).unwrap().eq(&filter))
            .map(|bit| String::from(bit))
            .collect();
        i += 1;
    }

    isize::from_str_radix(result.get(0).unwrap(), 2).unwrap()
}

File: golang/advent-of-code-2021/three/three.go
package three

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func ReadBitsFromFile() []string {
	file, err := os.Open("three/three.txt")
	if err != nil {
		panic("Cannot find/open input file for day 3")
	}

	reader := bufio.NewReader(file)
	result := make([]string, 0)
	for {
		currentLine, _, err := reader.ReadLine()
		if err != nil {
			break
		}
		result = append(result, string(currentLine))
	}
	return result
}

func CalculatePowerConsumption(bits []string) int {
	bitCounter := make([]int, len(bits[0]))
	for _, bitLine := range bits {
		for i, character := range bitLine {
			if character == '0' {
				bitCounter[i] -= 1
			} else {
				bitCounter[i] += 1
			}
		}
	}

	gamma := ""
	epsilon := ""
	for _, count := range bitCounter {
		if count >= 0 {
			gamma += "1"
			epsilon += "0"
		} else {
			gamma += "0"
			epsilon += "1"
		}
	}

	gammaValue, err := strconv.ParseInt(gamma, 2, 0)
	if err != nil {
		panic(fmt.Sprintf("Cannot parse %s into int with err %v", gamma, err))
	}

	epsilonValue, err := strconv.ParseInt(epsilon, 2, 0)
	if err != nil {
		panic(fmt.Sprintf("Cannot parse %s into int with err %v", gamma, err))
	}

	return int(gammaValue) * int(epsilonValue)
}

func CalculateOxygenRating(bits []string) int {
	index := 0

	for {
		if len(bits) <= 1 {
			break
		}

		count := 0
		for _, bitLine := range bits {
			if rune(bitLine[index]) == '0' {
				count--
			} else {
				count++
			}
		}

		if count >= 0 {
			bits = filterBits(bits, '1', index)
		} else {
			bits = filterBits(bits, '0', index)
		}

		index++
	}

	rating, err := strconv.ParseInt(bits[0], 2, 0)
	if err != nil {
		panic(fmt.Sprintf("Cannot parse %s into int with err %v", bits[0], err))
	}

	return int(rating)
}

func CalculateCO2Rating(bits []string) int {
	index := 0

	for {
		if len(bits) <= 1 {
			break
		}

		count := 0
		for _, bitLine := range bits {
			if rune(bitLine[index]) == '0' {
				count--
			} else {
				count++
			}
		}

		if count >= 0 {
			bits = filterBits(bits, '0', index)
		} else {
			bits = filterBits(bits, '1', index)
		}

		index++
	}

	rating, err := strconv.ParseInt(bits[0], 2, 0)
	if err != nil {
		panic(fmt.Sprintf("Cannot parse %s into int with err %v", bits[0], err))
	}

	return int(rating)
}

func filterBits(bits []string, filter rune, filterIndex int) []string {
	result := make([]string, 0)
	for _, bitLine := range bits {
		if bitLine[filterIndex] == byte(filter) {
			result = append(result, bitLine)
		}
	}

	return result
}
